docs/proposals/custom-metrics.md

```diff
@@ -0,0 +1,119 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/custom-metrics.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Custom Metrics in Kubernetes
+
+
+## Preface
+
+Our aim is to create a mechanism in Kubernetes that will allow pods to expose custom metrics, collect them, and make them accessible.
+Custom metrics are needed by:
+* horizontal pod autoscaler, for autoscaling the number of pods based on them;
+* scheduler, for using them in a sophisticated scheduling algorithm;
+* general monitoring of application health by a user (we are not sure yet whether this should be handled separately, by a different,
+general monitoring solution, see [#665](https://github.com/kubernetes/heapster/issues/665));
+* ...
+
+
+## Design
+
+For the Kubernetes version 1.2, we plan to implement aggregation of pod custom metrics in Prometheus format by pull.
+
+Each pod, to expose custom metrics, will expose a set of Prometheus endpoints.
+CAdvisor will collect metrics from such endpoints of all pods on each node by pulling, and expose them to Heapster.
+Heapster will:
+* collect custom metrics from all CAdvisors in the cluster, together with pulling system metrics
+(for version 1.2: we assume pooling period of ~30 seconds and the number of custom metrics ~3 per pod,
+we are not going to enforce the number by any quota),
```

Comment:
It might be a good idea, at least for the first iteration, to just have a tunable option on the kubelet called "max custom metrics" (per container) or something of the sort.  This would provide an initial way for cluster owners to limit the number of custom metrics while we figure out better controls.